### PR TITLE
Etcd Client Improve

### PR DESCRIPTION
**Pull request by ztelur**
**What this PR does**:

- fix create and update Compare error, use CreateRevision and ModRevision instead of Version
- return ErrCompareFail  when create and update txn compare fail
- provide bulk insertion API for batch insert
- provide getValAndRev API which can be used to get and watch with revision for pixiu admin and dubbo-go
- provide watch with OpOption API for watch with revision
- provide NewConfigClientWithErr to return client init error

### CODE REVIEW

**Review by AlexStocks — APPROVED** [review]

File: database/kv/etcd/v3/client.go [hunk 1]
```diff
@@ -211,35 +234,98 @@ func (c *Client) GetEndPoints() []string {
 	return c.endpoints
 }
 
-// if k not exist will put k/v in etcd, otherwise return nil
-func (c *Client) put(k string, v string, opts ...clientv3.OpOption) error {
+// if k not exist will put k/v in etcd, otherwise return ErrCompareFail
+func (c *Client) create(k string, v string, opts ...clientv3.OpOption) error {
 	rawClient := c.GetRawClient()
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:240]: delete line

File: database/kv/etcd/v3/client.go [hunk 2]
```diff
@@ -211,35 +234,98 @@ func (c *Client) GetEndPoints() []string {
...
+	if !resp.Succeeded {
+		return ErrCompareFail
+	}
+	return nil
+}
+
+// if k in bulk insertion not exist all, then put all k/v in etcd, otherwise return error
+func (c *Client) batchCreate(kList []string, vList []string, opts ...clientv3.OpOption) error {
+	rawClient := c.GetRawClient()
+
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:262]: delete this line

File: database/kv/etcd/v3/client.go [hunk 3]
```diff
@@ -211,35 +234,98 @@ func (c *Client) GetEndPoints() []string {
...
+	for i, k := range kList {
+		v := vList[i]
+		cs = append(cs, clientv3.Compare(clientv3.CreateRevision(k), "=", 0))
+		ops = append(ops, clientv3.OpPut(k, v, opts...))
+	}
+
+	resp, err := rawClient.Txn(c.ctx).
+		If(cs...).
+		Then(ops...).
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:284]: may be put these a line, seems better

File: database/kv/etcd/v3/client.go [hunk 4]
```diff
@@ -211,35 +234,98 @@ func (c *Client) GetEndPoints() []string {
...
+	if !resp.Succeeded {
+		return ErrCompareFail
+	}
+	return nil
+}
+
+// put k v into etcd
+func (c *Client) put(k string, v string, opts ...clientv3.OpOption) error {
+	rawClient := c.GetRawClient()
+
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:299]: delete the line

File: database/kv/etcd/v3/client.go [hunk 5]
```diff
@@ -211,35 +234,98 @@ func (c *Client) GetEndPoints() []string {
 	return c.endpoints
 }
 
-// if k not exist will put k/v in etcd, otherwise return nil
-func (c *Client) put(k string, v string, opts ...clientv3.OpOption) error {
+// if k not exist will put k/v in etcd, otherwise return ErrCompareFail
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:237]: comment is like this, is better. 

// put do xxxx. if k not exist xxxx

File: database/kv/etcd/v3/client.go [hunk 6]
```diff
@@ -272,6 +358,26 @@ func (c *Client) get(k string) (string, error) {
 	return string(resp.Kvs[0].Value), nil
 }
 
+// getValAndRev get value and revision
+func (c *Client) getValAndRev(k string) (string, int64, error) {
+	rawClient := c.GetRawClient()
+
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:364]: delete line

File: database/kv/etcd/v3/client.go [hunk 7]
```diff
@@ -310,24 +416,15 @@ func (c *Client) GetChildren(k string) ([]string, []string, error) {
 	return kList, vList, nil
 }
 
-func (c *Client) watchWithPrefix(prefix string) (clientv3.WatchChan, error) {
+// watchWithOption watch
+func (c *Client) watchWithOption(k string, opts ...clientv3.OpOption) (clientv3.WatchChan, error) {
 	rawClient := c.GetRawClient()
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:422]: delete this line

File: database/kv/etcd/v3/client.go [hunk 8]
```diff
@@ -211,35 +234,98 @@ func (c *Client) GetEndPoints() []string {
...
+	_, err := rawClient.Put(c.ctx, k, v, opts...)
 	return err
 }
 
 // if k not exist will put k/v in etcd
-// if k is already exist in etcd, replace it
-func (c *Client) update(k string, v string, opts ...clientv3.OpOption) error {
+// if k is already exist in etcd and revision = revision, replace it, otherwise return ErrCompareFail
+func (c *Client) updateWithRev(k string, v string, rev int64, opts ...clientv3.OpOption) error {
 	rawClient := c.GetRawClient()
```
Comment (georgehao) [inline database/kv/etcd/v3/client.go:312]: delete line

File: database/kv/etcd/v3/client.go (hunk 1 above)
Comment (ztelur) [reply database/kv/etcd/v3/client.go:240]: delete the comment?

Comment (ztelur) [reply database/kv/etcd/v3/client.go:240]: > delete line

delete the comment or blank line?

File: database/kv/etcd/v3/client.go [hunk 9]
```diff
@@ -36,10 +37,26 @@ var (
 	ErrNilETCDV3Client = perrors.New("etcd raw client is nil") // full describe the ERR
 	// ErrKVPairNotFound not found key
 	ErrKVPairNotFound = perrors.New("k/v pair not found")
+	// ErrKVListSizeIllegal k/v list empty or not equal size
+	ErrKVListSizeIllegal = perrors.New("k/v List is empty or kList size not equal to vList size")
```
Comment (AlexStocks) [inline database/kv/etcd/v3/client.go:41]: k/v List is empty or kList size is not equal to vList size

File: database/kv/etcd/v3/client.go [hunk 10]
```diff
@@ -36,10 +37,26 @@ var (
...
+	// ErrCompareFail txn compare fail
+	ErrCompareFail = perrors.New("txn compare fail")
+	// ErrRevision revision when error
+	ErrRevision int64 = -1
 )
 
 // NewConfigClient create new Client
 func NewConfigClient(opts ...Option) *Client {
+	newClient, err := NewConfigClientWithErr(opts...)
+
```
Comment (AlexStocks) [inline database/kv/etcd/v3/client.go:51]: delete this blank line.

File: database/kv/etcd/v3/client.go (hunk 3 above)
Comment (AlexStocks) [reply database/kv/etcd/v3/client.go:276]: @georgehao I do not think it is a problem.

File: database/kv/etcd/v3/client.go [hunk 11]
```diff
@@ -211,40 +228,97 @@ func (c *Client) GetEndPoints() []string {
...
+	if rawClient == nil {
+		return ErrNilETCDV3Client
+	}
+
+	resp, err := rawClient.Txn(c.ctx).
+		If(clientv3.Compare(clientv3.ModRevision(k), "=", rev)).
 		Then(clientv3.OpPut(k, v, opts...)).
 		Commit()
-	return err
+
```
Comment (AlexStocks) [inline database/kv/etcd/v3/client.go:310]: pld do not put a blank line btw the function and its 'error check if' clause.

**Review by zhaoyunxing92 — APPROVED** [review]
approved these changes

**Review by LaurenceLiZhixin — APPROVED** [review]

**Pull request merged**